File: app/src/main/java/com/snaptiongame/app/data/providers/LeaderboardProvider.java
package com.snaptiongame.app.data.providers;

import com.snaptiongame.app.data.api.SnaptionApi;
import com.snaptiongame.app.data.models.User;
import com.snaptiongame.app.data.providers.api.ApiProvider;

import java.util.ArrayList;
import java.util.List;

import io.reactivex.Single;

/**
 * @author Tyler Wong
 */

public class LeaderboardProvider {
    private static SnaptionApi apiService = ApiProvider.getApiService();

    public static Single<List<User>> getExperienceLeaderboard() {
        return Single.just(new ArrayList<>());
    }
}

File: app/src/main/java/com/snaptiongame/app/data/providers/LeaderboardProvider.kt
@file:JvmName("LeaderboardProvider")

package com.snaptiongame.app.data.providers

import com.snaptiongame.app.data.models.Friend
import com.snaptiongame.app.data.providers.api.ApiProvider

import io.reactivex.Single

/**
 * @author Tyler Wong
 */

private val apiService = ApiProvider.getApiService()

fun getUserLeaderboard(friendsOnly: Boolean): Single<List<Friend>> =
    apiService.getUserLeaderboard(friendsOnly)
            .flatMapIterable { user -> user }
            .map { user -> Friend(user) }
            .toList()
